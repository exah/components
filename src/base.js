import React from 'react'
import { identity } from '@exah/utils'

export function base ({
  use: defaultComp = 'div',
  filter = identity
} = {}) {
  const Base = React.forwardRef(({ as = defaultComp, use: Comp = as, ...rest }, ref) => (
    <Comp ref={ref} {...filter(rest, Comp)} />
  ))

  Base.displayName = 'Base'
  return Base
}
