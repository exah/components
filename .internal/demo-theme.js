import { css } from '@emotion/core'

const Link = css`
  text-decoration: none;

  &:hover {
    text-decoration: underline;
  }
`

const Button = css`
  border-radius: 5px;
  padding: 10px 20px;
`

const Input = css`
  padding: 10px 20px;
`

export const theme = {
  media: {
    sm: '(max-width: 600px)',
    md: '(min-width: 601px) and (max-width: 1024px)',
    lg: '(min-width: 1025px)'
  },
  space: [ 0, 8, 16, 32, 64, 128 ],
  size: {
    large: 128,
    medium: 56,
    small: 42
  },
  color: {
    black: '#000',
    grey: [ '#eee', '#ddd', '#ccc', '#bbb', '#aaa', '#555' ],
    white: '#fff'
  },
  palette: {
    default: {
      bg: '#fff',
      fg: '#000',
      link: 'royalblue',
      primary: '#ff0',
      positive: '#0f0',
      negative: '#f00'
    },
    grey: {
      bg: '#eee',
      fg: '#000'
    },
    inverted: {
      bg: '#000',
      fg: '#fff'
    }
  },
  fontFamily: {
    default: 'system-ui, sans-serif',
    'system-ui': [
      '-apple-system',
      'BlinkMacSystemFont',
      'Segoe UI',
      'Roboto',
      'Oxygen-Sans',
      'Ubuntu',
      'Cantarell',
      'Helvetica Neue',
      'sans-serif',
    ].join(),
    serif: 'Times New Roman, serif'
  },
  textStyle: {
    root: {
      fontFamily: 'Helvetica, system-ui, sans-serif',
      fontSize: 16,
      lineHeight: 1.2
    },
    default: {
      fontFamily: 'Helvetica, system-ui, sans-serif',
      fontSize: '1rem',
      lineHeight: 1.2
    },
    heading: {
      fontFamily: 'athelas, georgia, times, serif',
      fontSize: '2.5rem',
      fontWeight: 'bold'
    },
    responsive: {
      all: {
        fontSize: '1rem'
      },
      lg: {
        fontSize: '2rem'
      },
      md: {
        fontSize: '1.5rem'
      }
    }
  },
  Link,
  Button,
  buttonStyle: {
    default: {
      background: 'royalblue',
      color: 'white'
    },
    danger: {
      background: 'red',
      color: 'white'
    },
    success: {
      background: 'lime',
      color: 'white'
    },
    inverted: {
      background: 'white',
      color: 'black',
    }
  },
  Input,
  inputStyle: {
    default: {
      color: 'white',
      background: '#2e3440',
      border: '1px solid rgba(255, 255, 255, 0.3)',
    },
    underline: {
      color: 'white',
      padding: '10px 0',
      borderBottom: '1px solid rgba(255, 255, 255, 0.3)',
    },
    inverted: {
      color: 'black',
      background: 'white',
      border: '1px solid lime',
    },
  },
  gridStyle: {
    'default': {
      gridTemplateColumns: 'repeat(12, 1fr)',
      gridGap: '1rem'
    },
    '16cols': {
      gridTemplateColumns: 'repeat(16, 1fr)',
      gridGap: '0.5rem'
    },
    'page-layout': {
      gridTemplate: `
       "head head" 30px
       "nav  main" 1fr
       "nav  foot" 30px / 120px 1fr
     `
    }
  }
}
