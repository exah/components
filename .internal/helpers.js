import renderer from 'react-test-renderer'
import expect from 'expect'

export const renderJSON = (element) =>
  renderer.create(element).toJSON()

export const anyClassName = expect.stringMatching(/^\./)
