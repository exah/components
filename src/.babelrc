{
  "env": {
    "modules": {
      "plugins": [
        "transform-es2015-modules-commonjs",
        [ "emotion", { "hoist": true } ]
      ]
    },
    "production": {
      "plugins": [
        [ "emotion", { "hoist": true } ]
      ]
    },
    "development": {
      "plugins": [
        ["emotion", { "sourceMap": true }]
      ]
    }
  },
  "presets": [
    [ "env", {
      "useBuiltIns": true,
      "modules": false,
      "loose": true
    } ]
  ],
  "plugins": [
    "transform-react-jsx",
    "transform-class-properties",
    [ "transform-object-rest-spread", { "useBuiltIns": true } ]
  ]
}
