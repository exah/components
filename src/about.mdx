---
name: Getting Started
route: /
order: 1
---

# Getting Started
