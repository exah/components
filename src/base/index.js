import Base, { createBase } from 'react-styled-base'

export {
  Base,
  createBase
}
