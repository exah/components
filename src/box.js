import styled from '@emotion/styled'
import { getPropTypes } from 'pss/prop-type'
import { combineStyles, box, boxStyle, themePath } from 'pss'
import { base, omit } from './utils'

const styles = combineStyles(
  boxStyle,
  box
)

const boxDefaultStyles = {
  font: 'inherit',
  color: 'inherit',
  background: 'transparent',
  border: 0,
  margin: 0,
  padding: 0
}

const BaseBox = base({
  name: 'Box',
  filter: omit(styles.props)
})

const Box = styled(BaseBox)(
  { boxSizing: 'border-box' },
  themePath('Box', boxDefaultStyles),
  styles
)

Box.displayName = 'Box'

Box.propTypes = {
  ...BaseBox.propTypes,
  ...getPropTypes(styles)
}

export {
  Box
}
