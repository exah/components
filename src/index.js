export { Box } from './box'
export { Button } from './button'
export { Feed } from './feed'
export { Flex } from './flex'
export { FlexGrid } from './flex-grid'
export { Grid } from './grid'
export { Input } from './input'
export { Image } from './image'
export { Layout } from './layout'
export { Link } from './link'
export { List } from './list'

export {
  MatchMediaContext,
  MatchMediaProvider,
  MatchMediaConsumer,
  useMatchMedia,
  useMatchMediaContext,
  withMatchMedia
} from './match-media'

export { Text } from './text'

export {
  ThemeDefaults,
  ThemeProvider,
  useTheme
} from './theme'

export { base } from './utils'
