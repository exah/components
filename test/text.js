import React from 'react'
import { renderJSON, anyClassName } from '../.internal/helpers'
import { theme } from '../.internal/demo-theme'
import { Text, ThemeProvider } from '../src'

describe('Text', () => {
  test('empty', () => {
    const result = renderJSON(<Text />)

    expect(result).toMatchSnapshot()
  })

  test('with text', () => {
    const result = renderJSON(<Text>Hello</Text>)

    expect(result).toMatchSnapshot()
  })

  test('with theme', () => {
    const result = renderJSON(
      <ThemeProvider theme={theme}>
        <Text variant='auto'>Hello</Text>
      </ThemeProvider>
    )

    expect(result).toMatchSnapshot()
  })

  test('as selector', () => {
    expect(`${Text}`).toEqual(anyClassName)
  })
})
