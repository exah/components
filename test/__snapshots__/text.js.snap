// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text empty 1`] = `
.emotion-1 {
  box-sizing: border-box;
  font: inherit;
  color: inherit;
  background: transparent;
  border: 0;
  margin: 0;
  padding: 0;
}

<div
  className="emotion-0 emotion-1 emotion-2"
/>
`;

exports[`Text with text 1`] = `
.emotion-1 {
  box-sizing: border-box;
  font: inherit;
  color: inherit;
  background: transparent;
  border: 0;
  margin: 0;
  padding: 0;
}

<div
  className="emotion-0 emotion-1 emotion-2"
>
  Hello
</div>
`;

exports[`Text with theme 1`] = `
.emotion-1 {
  box-sizing: border-box;
  font: inherit;
  color: inherit;
  background: transparent;
  border: 0;
  margin: 0;
  padding: 0;
  font-family: Helvetica,system-ui,sans-serif;
  font-size: 1rem;
  line-height: 1.2;
}

<div
  className="emotion-0 emotion-1 emotion-2"
>
  Hello
</div>
`;
